### ISSUE TITLE
Don't allow `as` to be used in casting, it does more harm than good

### ISSUE THREAD

**Issue by dark-chocolate** [dart-lang/linter#2102]
Referring [this thread](https://github.com/dart-lang/pedantic/issues/55) from pedantic.

You can either use `as` or type annotation to upcast or downcast things, let's take this example, where I have a `SizedBox` and I want to upcast to a `Widget`, so that **I don't want to allow users to use `width` property** on it, in this example I am using `as` to upcast, but this gives me a warning:

_(Whenever I am writing `bad` in comment, that means it is undesired behavior, and `good` for desired one)_

```
const widget = SizedBox() as Widget; // warning 
final width = widget.width; // error (good)
```
Now if I take warning into consideration and modify my code to silence pedantic, I can do:
```
const widget = SizedBox(); // no warning (good)
final width = widget.width; // no error (bad)
```

But this allows me to access `width` which I don't want. 

Second solution is to use type annotation like:
```    
const Widget widget = SizedBox(); // no warning (good)
final width = widget.width; // error (good)
```

So you can clearly see if I use `as`, I get myself in trouble and can only do limited things but if I use type annotation, I can do things the way I want. Let's take one more example into consideration (thanks to @eernstg )
```
final foo = 1 as String; // no error (bad)
final String bar = 1; // error (good)
```

As you can see if I am again comparing `as` with type annotation, the latter has advantages almost everywhere.

**Comment by bwilkerson** [comment]
While I agree with what you're saying, I don't understand what you're asking for in terms of the tooling. There's already a notification if you attempt to upcast, so I'm not sure what else we can do. (Other than adding a quick fix that both removes the cast and adds a type annotation.)

**Comment by dark-chocolate** [comment]
@bwilkerson  I am just requesting Dart team to either silence the linter when I am using `as` or simply remove `as` from Dart (which I don't understand servers any purpose that can't be possible with type annotation as I mentioned in the example, it's all on the basis of my limited knowledge of Dart, so there might be some cases where `as` comes to rescue where type annotation fails). 

Hope I was clear!

**Comment by dark-chocolate** [comment]
Ok, so I am able to create an example where as works but type annotation doesn't (in terms of warning)
```
final list = <Object>[1];
final listInts = list.cast() as List<int>; // no warning (good)
final List<int> listInts = list.cast(); // warning (bad)
```

Can we stick to one design pattern, I request team to just silence `as` whenever it is used, or simply allow cases where `List<int>` can be used in type annotation (above one shows error, it shouldn't).

**Comment by eernstg** [comment]
It is not likely to be a good idea to remove `as` from the language entirely: it would surely be a seriously breaking change, and there are cases where you need it. For instance:

```dart
class A {}
class B {}
class C implements A, B {}

void main() {
  A a = C();
  if (a is B) {
    B b = a as B; // Promotion not applicable, must declare `b`.
    ...
  }
}
```

In order to keep type promotion comprehensible it will only "promote" the type of a local variable in a certain scope to a subtype, it won't change the type arbitrarily.

So even though you may wish to protect all your cast-like situations by an `is` check, you still can't obtain a typing of `a` as a `B`. So you need `a as B` in order to obtain a reference to the given object typed as `B`. Moreover, you may not be able to test for `C`, because there could be many types implementing both `A` and `B`. In general: For any two unrelated types `S` and `T`, there could more types implementing both than the ones that you know about. Of course, if it is always a bug when the cast does not succeed, you may wish to avoid the `is` test, and then you still need `as`.

With that in mind, it isn't obvious how to deal with `as`. As mentioned, it is a bad idea to use a cast as a replacement for a type annotation:

```dart
void main() {
  var x1 = SomeWidget(...) as Widget; // Dangerous! ..
  Widget x2 = SomeWidget(...); // .. if it is just intended to mean this.
}
```

But it is not obvious to me that we can easily make the distinction between the situations where the cast is justified and when it is not. Even the most deeply justified ones (like casting to a completely different type, based on carefully maintained subtype relations and invariants) look very much like the plain too-many-thumbs bug.

One thing you _could_ do, however, is to enable the lint [`avoid_as`](https://dart-lang.github.io/linter/lints/avoid_as.html), and manually add an `// ignore: avoid_as` on every cast which is actually carefully justified.

**Comment by dark-chocolate** [comment]
@eernstg  Taking this example 

```
void main() {
  A a = C();
  if (a is B) {
   B b = a as B; // should there be any need of doing this because a is already B?
   // I suppose this should be done implicitly so that I can access the properties of `B`, above line should be redundant. 
  }

  Object o = 'o';
  if (o is String) {
    o.contains('o'); // I don't need to cast this to a type String because o is already String
  }
}
```

I know all classes extends `Object` so I don't need `as` in second example but can we actually do the same kind of things in other places too (a feature request maybe?)

**Comment by bwilkerson** [comment]
> ... silence the linter when I am using `as` ...

Sorry, but I need something more specific. Can you specify which lint(s) you're asking to have changed and under what conditions?

**Comment by dark-chocolate** [comment]
@bwilkerson For now, I can only think of a case when we are upcasting, like this example (which we've been discussing so far)

```
const widget = SizedBox() as Widget; // warning: unnecessary_cast
```

There shouldn't be a warning because I don't want to use `SizedBox` properties on `widget`.  To make it work I'll need to modify my linter rules. So, I am not sure how big this context could be when you want to silence the linter.

**Comment by bwilkerson** [comment]
I'm confused. I thought you had decided that you don't want to use `as` in that case, but that you thought it was better to use a type annotation. Hence the warning is appropriate.

**Comment by dark-chocolate** [comment]
Even I am confused, I understand we have been jumping around a lot from this thread to other and vice-versa. I am giving you this final example to demonstrate what I actually want to complain about:

```    
// They are doing the same thing (according to dart linter)
const Widget widget1 = SizedBox(); // no warning
const widget2 = SizedBox(); // no warning

// Getting access to [width] property on different instances
final width1 = widget1.width; // error (desired behavior)
final width2 = widget2.width; // no error (undesired behavior)

// In order to fix [width2] to produce error, I'll have to upcast [widget2] using `as`
const widget2Fixed = SizedBox() as Widget; // but this gives me warning not to do so
final width2Fixed = widget2Fixed.width; // no error (desired behavior)
```

**Comment by bwilkerson** [comment]
```
// They are doing the same thing (according to dart linter)
const Widget widget1 = SizedBox(); // no warning
const widget2 = SizedBox(); // no warning
```

There are _not_ doing the same thing. The fact that the linter isn't producing any lints only indicates that neither one violates any of the enabled lint rules. It doesn't mean that they have the same semantics, nor that there aren't any bugs.

The first line declares a variable named `widget1` whose static type is `Widget` and initializes that variable to an instance of the class `SizedBox`. Later on, when you attempt to access the member `width`, you get the error you desire because the class `Widget`, the static type of the variable, does not declare a member named `width`.

The second line declares a variable named `widget2` whose static type is `SizedBox` and initializes that variable to an instance of the class `SizedBox`. (The reason the static type is `SizedBox` is because it has no explicit type annotation and hence the type was inferred from the initializer.) Later on, when you attempt to access the member `width`, you do not get the error you desire because the class `SizedBox`, the static type of the variable, _does_ declare a member named `width`.

The reason that the example `widget2Fixed` works is because the `as` expression changes the static type of the initialize expression to be `Widget`. But as you pointed out earlier in the conversation, it's better to use an explicit type annotation than an `as` expression because the type annotation can be used an analysis time while the type cast is a runtime check that isn't checked at analysis time.

**Comment by dark-chocolate** [comment]
@bwilkerson Thank you sir, the 1st paragraph cleared up things. And I hope you were not taunting me in rest of your article because I do understand those basic concept. I hope this kind of concepts (difference between type annotation and `as`, which one to use when) are also mentioned somewhere by Dart team in their documentation. 

And I really appreciate your time clearing all my doubts with extreme patience and humanity (You guys deserve oscars, how humble you were, no matter how stupid the question was)

I am closing this issue now and want see you again on some other issues (and if you don't want to see me, you can share the secret source of your Dart knowledge ;)

**Comment by bwilkerson** [comment]
> And I hope you were not taunting me in rest of your article ...

Absolutely not! I'm sorry if it came across that way! I was merely being complete in an attempt to clear up the confusion as quickly as possible.

> I hope this kind of concepts (difference between type annotation and `as`, which one to use when) are also mentioned somewhere by Dart team in their documentation.

@kwalrath Some feedback on concepts that might need to be explained (they probably already are, but I'm not familiar enough with the docs to have been able to point to them for the purpose of this discussion).

**Comment by dark-chocolate** [comment]
@bwilkerson 
> Absolutely not! I'm sorry if it came across that way! I was merely being complete in an attempt to clear up the confusion as quickly as possible.

No sir, actually I am sorry for misinterpreting the way you were explaining. Thanks for clearing up everything :)
